modules/openbsd/openbsd-driver.c

```diff
@@ -0,0 +1,135 @@
+#include <sys/ioctl.h>
+#include <fcntl.h>
+#include <errno.h>
+#include <unistd.h>
+
+#include "openbsd-driver.h"
+
+#include "logsource.h"
+#include "messages.h"
+#include "logreader.h"
+#include "fdhelpers.h"
+#include "apphook.h"
+#include "stats/stats-registry.h"
+#include "poll-fd-events.h"
+#include "logproto/logproto-dgram-server.h"
+#include "transport/transport-socket.h"
+
+#ifndef LIOCSFD
+#define LIOCSFD   _IOW('l', 127, int) /* set sendsyslog() fd */
+#endif
+
+typedef struct _openBSDDriver
+{
+  LogSrcDriver     super;
+  LogReaderOptions reader_options;
+  LogReader        *reader;
+  GString          *dev_filename;
+  int              klog;
+  int              pair[2];
+} OpenBSDDriver;
+
+static void
+_openbsd_sd_free(LogPipe *s)
+{
+  OpenBSDDriver *self = (OpenBSDDriver *)s;
+  log_reader_options_destroy(&self->reader_options);
+  log_src_driver_free(s);
+}
+
+static gboolean
+_openbsd_sd_init(LogPipe *s)
+{
+  OpenBSDDriver *self = (OpenBSDDriver *)s;
+  GlobalConfig *cfg = log_pipe_get_config(s);
+
+  if (!log_src_driver_init_method(s))
+    return FALSE;
+
+  log_reader_options_init(&self->reader_options, cfg, self->super.super.group);
+
+  if (socketpair(AF_UNIX, SOCK_STREAM, PF_UNSPEC, self->pair) == -1)
```

Comment:
Maybe the openBSD specific part could be extracted to a function that returns only with an _fd_ that the driver should watch. Something like `int openbsd_syslog_dev_open()` and possible the counterpart `void openbsd_syslog_dev_close()`.

Reply:
As the socket open is used only in one place, I don't see the benefit of creating a dedicated function for it.

Reply:
The benefit would be more readable code, because the creation of socket pair is one thing to do. Different task we can wrap up with functions even if used only once.
The `openbsd_sd_init` would be something like: create an openbsd newsyslog socket and watch it with dgram server and stream transport. 

Instead now something like: createa a socket pair, open `/dev/klog` send some magical stuff to the `/dev/klog`, close part of the socket pair, ...

The code could be like this:
```c
static int openbsd_create_newsyslog_socket(OpenBSDDriver *s)
{
   int pair[2];
  ...
  close(pair[1]);
  return pair[0];
}

static gboolean
_openbsd_sd_init(LogPipe *s)
{
  OpenBSDDriver *self = (OpenBSDDriver *)s;
  GlobalConfig *cfg = log_pipe_get_config(s);
...
  int fd = openbsd_create_newsyslog_socket(self);

  log_reader_reopen(self->reader, log_proto_dgram_server_new(log_transport_stream_socket_new(fd),
                                                             &self->reader_options.proto_options.super),
                                                            poll_fd_events_new(fd));
...
```

I am okay with the current version also, because it is not bad. But I wanted to note, that in fact the request was not related only to remove any kind of duplication, although with additional refactor it would be possible. That is out of scope now I think.